apps/dhparam.c

```diff
@@ -191,221 +176,206 @@ int dhparam_main(int argc, char **argv)
+        if (!EVP_PKEY_paramgen_init(ctx)) {
             BIO_printf(bio_err,
-                       "Generating DH parameters, %d bit long safe prime, generator %d\n",
-                       num, g);
-            BIO_printf(bio_err, "This is going to take a long time\n");
-            if (!EVP_PKEY_paramgen_init(ctx)) {
-                BIO_printf(bio_err,
-                           "Error, unable to initialise DH param generation\n");
+                        "Error, unable to initialise %s parameters\n",
+                        alg);
+            goto end;
+        }
+
+        if (dsaparam) {
+            if (!EVP_PKEY_CTX_set_dsa_paramgen_bits(ctx, num)) {
+                BIO_printf(bio_err, "Error, unable to set DSA prime length\n");
                 goto end;
             }
-
+        } else {
             if (!EVP_PKEY_CTX_set_dh_paramgen_prime_len(ctx, num)) {
                 BIO_printf(bio_err, "Error, unable to set DH prime length\n");
                 goto end;
             }
-            if (!EVP_PKEY_paramgen(ctx, &pkey)) {
-                BIO_printf(bio_err, "Error, DH generation failed\n");
+        }
+
+        if (!EVP_PKEY_paramgen(ctx, &tmppkey)) {
+            BIO_printf(bio_err, "Error, %s generation failed\n", alg);
+            goto end;
+        }
+
+        EVP_PKEY_CTX_free(ctx);
+        ctx = NULL;
+        if (dsaparam) {
+            pkey = dsa_to_dh(tmppkey);
+            if (pkey == NULL)
                 goto end;
-            }
-            dh = EVP_PKEY_get0_DH(pkey);
+            EVP_PKEY_free(tmppkey);
+        } else {
+            pkey = tmppkey;
         }
+        tmppkey = NULL;
     } else {
+        OSSL_DECODER_CTX *decoderctx = NULL;
+
         in = bio_open_default(infile, 'r', informat);
         if (in == NULL)
             goto end;
 
-#if !defined(OPENSSL_NO_DSA) && !defined(OPENSSL_NO_DEPRECATED_3_0)
-        if (dsaparam) {
-            DSA *dsa;
-
-            if (informat == FORMAT_ASN1)
-                dsa = d2i_DSAparams_bio(in, NULL);
-            else                /* informat == FORMAT_PEM */
-                dsa = PEM_read_bio_DSAparams(in, NULL, NULL, NULL);
+        /*
+         * We set NULL for the keytype to allow any key type. We don't know
+         * if we're going to get DH or DHX (or DSA in the event of dsaparam).
+         * We check that we got one of those key types afterwards.
+         */
+        decoderctx
+            = OSSL_DECODER_CTX_new_by_EVP_PKEY(&tmppkey,
+                                                (informat == FORMAT_ASN1)
+                                                ? "DER" : "PEM", NULL,
+                                                NULL, NULL);
+        if (decoderctx == NULL
+                || !OSSL_DECODER_from_bio(decoderctx, in)) {
+            OSSL_DECODER_CTX_free(decoderctx);
+            BIO_printf(bio_err, "Error, unable to load parameters\n");
+            goto end;
+        }
+        OSSL_DECODER_CTX_free(decoderctx);
```

Comment:
In many other commands, we have dropped support for `-informat` (basically ignoring it for anything other than "engine", where that's interesting), and use `load_key()` and `load_pubkey()`.  So an alternative would be to implement `load_keyparams()` and use that instead.  All you'd have to do after calling that is a check that `EVP_PKEY_is_a(tmppkey, "DH")` returns true.

Reply:
(I have to remember to document that when I work on the decoders)

Reply:
For an example, see #13191

Reply:
Unfortunately its not that simple. I need different processing for DER format than I do PEM format. The reason is that I cannot leave the decoder to auto-detect what is in the file for DER. If I do it sometimes gets it wrong and thinks that a DHX file is DSA file. Therefore I have to explicitly tell it to try and load DH, and then if that fails do it for DHX....but only for DER. Its a bit horrible.